src/Interpreters/RewriteSumFunctionWithSumAndCountVisitor.cpp

```diff
@@ -0,0 +1,125 @@
+#include <Interpreters/RewriteSumFunctionWithSumAndCountVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+
+namespace DB
+{
+
+void RewriteSumFunctionWithSumAndCountMatcher::visit(ASTPtr & ast, const Data & data)
+{
+    if (auto * func = ast->as<ASTFunction>())
+        visit(*func, ast, data);
+}
+
+/** Rewrite the following AST to break the function `sum(column +/- literal)` into two individual functions
+ * `sum(column)` and `literal * count(column)`.
+ * sum(column + literal) -> sum(column) + literal * count(column)
+ * sum(literal + column) -> literal * count(column) + sum(column)
+ * sum(column - literal) -> sum(column) - literal * count(column)
+ * sum(literal - column) -> literal * count(column) - sum(column)
+ */
+void RewriteSumFunctionWithSumAndCountMatcher::visit(const ASTFunction & function, ASTPtr & ast, const Data & data)
+{
+    static const std::unordered_set<String> function_supported = {
+        "plus",
+        "minus"
+    };
+
+    if (!function.arguments || Poco::toLower(function.name) != "sum" || function.arguments->children.size() != 1)
+        return;
+
+    const auto * func_plus_minus = function.arguments->children[0]->as<ASTFunction>();
+
+    if (!func_plus_minus || !function_supported.contains(Poco::toLower(func_plus_minus->name)) || func_plus_minus->arguments->children.size() != 2)
+        return;
+
+    size_t column_id = func_plus_minus->arguments->children.size();
+
+    for (size_t i = 0; i < func_plus_minus->arguments->children.size(); i++)
+    {
+        if (func_plus_minus->arguments->children[i]->as<ASTIdentifier>())
+            column_id = i;
+    }
+
+    if (column_id == func_plus_minus->arguments->children.size())
+        return;
+
+    size_t literal_id = 1 - column_id;
+    const auto * literal = func_plus_minus->arguments->children[literal_id]->as<ASTLiteral>();
+    if (!literal)
+        return;
+
+    Field::Types::Which literal_type = literal->value.getType();
```

Comment:
Leaving a comment here about why these types are accepted and not others would be nice

Reply:
The comment is added.